### ISSUE TITLE
Cannot connect: Object Reference not set to an instance of an object

### ISSUE THREAD

**Issue by JakobGSvendsen** [jpurgar/aa-scsm-connector#1]
Cannot connect. I am logging on as a organizational user on our ADFS. I have only 1 automation account that was created through ASM, which means it can be seen in both portals.

I have tried to pull the source from git, but I have a few issues. Main issue is that a file called HandleBase.cs is missing. Can you add that file to the repo?
(Besides that I have a problem in finding the refernced SM files, but that works on my colleagues PC so must be something on my PC. I have the SCSM console installed with UR9)

![image](https://cloud.githubusercontent.com/assets/3460807/14144221/14592faa-f68f-11e5-8a6b-7ded3985bcba.png)

P.s.
Great work! I like the design and it works pretty good on my colleagues PC on his account. We have been planning to create this kind of connector for a while but havent had the time. Would love to contribute if I can make the project work :)

**Comment by jurepurgar** [comment]
Hi,

I will have to check in detail what is going on but, it is possible that no automation accounts are visible to the client. Do you see any accounts if you use Get-AzureRMAutomationAccount cmdlet?

I've added HandlerBase.cs to the repo along with some other missing files :). I've also added MP references. SM dll references should be found if you have console installed. 

Let me know how it goes, and thanks for interest.

**Comment by JakobGSvendsen** [comment]
yes the accounts is listed with the cmdlet. Thanks for updating the project,. Will try to compile and debug where the object reference is mising.. Might m´be one of the values .. 

Here is the output maybe you can see whats wrong:

SubscriptionId        : 7b58a17c-0824-40c7-a792-6176c19f65a3
ResourceGroupName     : OaaSCS2WT6LFDWV7UA37577W3S7SWTQ5SJVJUDRBZHKSSWDXNA-West-Europe
AutomationAccountName : CoretechAutomation
Location              : West Europe
State                 :
Plan                  :
CreationTime          : 25-03-2015 21:13:54 +01:00
LastModifiedTime      : 29-12-2015 12:39:12 +01:00
LastModifiedBy        :
Tags                  : {}

**Comment by jurepurgar** [comment]
I have no idea what could be wrong :) Anyway I've published a new release with better error handling. Error message should now also show stack trace. I'll try to use SCSM error dialog in the future.
Just run the install script again and MP and dlls should be upgraded.

**Comment by JakobGSvendsen** [comment]
thanks for the update! .. I installed it and tested. but now I get no error message, it just reopens the logon dialog after it has used a few seconds in "searching for automation accounts". tried rebooting the scsm server to see if it helped, but no luck.

I see the missing files in the project now. but even when I have the console installed. most of the references are still gone (but as I mentioned before, it did work on my colleagues PC). 
![image](https://cloud.githubusercontent.com/assets/3460807/14168141/917447f4-f720-11e5-8ca0-3c150812a1fd.png)

Will try to fix it when I can. Maybe I helps if I install them into GAC. Which location do you have the SM console installed at?

**Comment by jurepurgar** [comment]
It might ask for login multiple times if refresh token does not work for multiple resources for some reason. What happens if you try to login if you are asked the second time?

Missing assemblies should be in C:\Program Files\Microsoft System Center 2012 R2\Service Manager. Maybe your console installation path is different? Do you think it would be a good idea to put assemblies in the repo so people will not have this issue?

**Comment by JakobGSvendsen** [comment]
ah stupid that i did not try twice, but cancelled and rebooted etc. LOL
It did work second time! :)
Now to test more stuff!. again. Great work! you have my full support, will blog about it as soon as I have time (I am fully booked today for some SCOM dev) 

and my console is in same path. so weird!
I can drag and drop the files on top of the failed ones to make them work, not sure if that changes anything if I later have a pull request. 
I think these problems with references is always the hard to get right.

I usually include them in my projects for customers/internal but I am not sure it is allowed to upload them in a open source project? If it is allowed I would recommend to put them in the projects and reference them via a relative path (it think it does that automatically if the files are in the project folder tree)

**Comment by JakobGSvendsen** [comment]
you/we do have some of the same problems in the MP project. The VSAE plugin sucks in many ways, this is one of them

![image](https://cloud.githubusercontent.com/assets/3460807/14172177/00d649ce-f737-11e5-8983-fed7dbd5dabe.png)

:) but it does work great now! .. I was only opening it to add a feature maybe (its great that you added the _parent item mappings, but I could really use a _root parent item too)

**Comment by JakobGSvendsen** [comment]
still having some problems in the sense of connecting.
1. It logged on after 2nd try and connected the account
2. I tried to create a activity, but the "select runbook" dialog keep searching forever (was there for 30 minutes while lunch).
3. I closed the console.
4. Reopened console
5. Tried to create a activity, this time it asked me for logon (also asked for the acceptance of the app permissions again as it does every time´).
6. then it was stuck and searching again.

Is it normal behavior that it asks for logon again after the console has been closed?

**Comment by jurepurgar** [comment]
I've added all referenced dlls to the repo, others have it too, so I think we should be fine. I was also able to solve the issue with MP references. I've manually updated .mpproj file with relative paths and it works. Let me know if you are now able to build the solution without hacking any references.

I'll try to write some error handling for "select runbook" tomorrow so we can see what is going on. 

It is normal that you are asked for creds every time you re-open the console since token cache is only in memory. I'm not sure why acceptance of the app is required every time, must be by design since it is the same for me. I'm not sure however, why you were asked for login twice while configuring connector. There are calls for management API and graph API which requires separate tokens, but it should be possible to get token by refresh token the second time. You can try to debug "AcquireTokenAsync" in ClientBase.cs to see what is going on. This is called for every api request so you can se what is going on with tokens.

I'm not sure what you meant by "_root parent item"? Can you open a separate issue and explain a bit more what the feature would be?

It would be great if you took the time to blog about it. I have to do it also :) and write more documentation :)

**Comment by JakobGSvendsen** [comment]
compilation worked great this time, except that they key is not in the solution. 
Maybe add the key file to a soltuion folder? (of cause not your own but maybe a special one for this project).

What I mean by root parent, is to be able to map to the ID of the service request that actually hosts the activity. The activity could be in several layers of activities (parallel/sequence), therefore the parent id might not be enough. Is the parent id property part of the normal set of properties on a activity? (I cant remember, and I thought you had made some code to actrually get the parent).
would be a great functionality, since that is how we worked before. but if it is not there , we just need a runbook to get the service request from the activity ID, which isnt hard to do either.

I haven't debugged yet because of too much work! but will soon. :)

**Comment by JakobGSvendsen** [comment]
btw. one of my customers also had to login twice, and has the problem on the runbook list.
Will debug asap

**Comment by JakobGSvendsen** [comment]
sorry for the slow feedback, been too damn busy. I have tried a few time to see why I need to logon 3 times, and why it does not show my runbooks, but I havent found the error yet. wanted to step through but it seems to be hard due to the async nature of the code (and that VS does not think I have all the symbols loaded for some reason)

**Comment by jurepurgar** [comment]
yes it is a bit hard to debug because of async. I usually put a break points after on the line after await so I can inspect the result of an asyn call. Can you maybe do a Fiddler trace, maybe I can figure out what is going on by inspecting http calls.

**Comment by jurepurgar** [comment]
and regarding root parent work item of automation activity: It actually already work that way. If automation activity is in service request you will always get ID of service request, regardless of multiple levels of activities (parallel/sequence). I have code for this, it's not a property of activity. And another thing: if runbook parameter type is object or object[] you can also get related and affected items, so there is no need to connect back to SM from runbook o get those.
